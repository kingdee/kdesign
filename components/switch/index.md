---
category: Components
type: 表单
order: 14
title: Switch
subtitle: 开关
---

开关是一种用于两种互斥状态之间进行切换的组件。

## 使用场景

需要触发激活或停用某些功能时，例如：“启用”和“不启用”。

## API

| 参数              | 说明                              | 类型                                     | 默认值 | 版本  |
| ----------------- | --------------------------------- | ---------------------------------------- | ------ | ----- |
| className         | Switch 器类名                     | string                                   | -      | 1.0.0 |
| checked           | 指定当前是否选中                  | boolean                                  | false  | 1.0.0 |
| checkedChildren   | 选中时的内容                      | string\|ReactNode                        |        | 1.0.0 |
| defaultChecked    | 初始是否选中                      | boolean                                  | false  | 1.0.0 |
| disabled          | 是否禁用                          | boolean                                  | false  | 1.0.0 |
| loading           | 加载中的开关                      | boolean                                  | false  | 1.0.0 |
| size              | 开关大小，可选值：`large` `small` | string                                   | small  | 1.0.0 |
| unCheckedChildren | 非选中时的内容                    | string\|ReactNode                        |        | 1.0.0 |
| onClick           | 点击时回调函数                    | Function(checked: boolean, event: Event) |        | 1.0.0 |
| onChange          | 变化时回调函数                    | Function(checked: boolean, event: Event) |        | 1.0.0 |

## Design Token

| 分类        | 组件 token                                       | 全局 token             | 默认值             |
| ----------- | ------------------------------------------------ | ---------------------- | ------------------ |
| color       | --kd-c-switch-color-on                           | --kd-g-color-theme-6   | rgb(85, 130, 243)  |
|             | --kd-c-switch-color-off                          | -                      | #B2B2B2            |
|             | --kd-c-switch-color-on-background-disabled       | --kd-g-color-theme-4   | rgb(181, 207, 255) |
|             | --kd-c-switch-color-off-background-disabled      | -                      | #D9D9D9            |
|             | --kd-c-switch-color-loading                      | --kd-g-color-ongoing   | #276ff5            |
|             | --kd-c-switch-color-font                         | --kd-g-color-white     | #fff               |
| font        | --kd-c-switch-small-font-size                    | --kd-g-font-size-small | 12px               |
|             | --kd-c-switch-large-font-size                    | --kd-g-font-size-small | 12px               |
|             | --kd-c-switch-inner-font-size-small              | --kd-g-font-size-small | 12px               |
| line-height | --kd-c-switch-large-line-height                  | -                      | 20px               |
|             | --kd-c-switch-small-line-height                  | -                      | 16px               |
| radius      | --kd-c-switch-small-radius-border                | -                      | 8px                |
| sizing      | --kd-c-switch-small-sizing-height                | -                      | 16px               |
|             | --kd-c-switch-small-sizing-width                 | -                      | 32px               |
|             | --kd-c-switch-large-sizing-height                | -                      | 20px               |
|             | --kd-c-switch-large-sizing-width                 | -                      | 40px               |
|             | --kd-c-switch-large-sizing-border-radius         | -                      | 10px               |
| spacing     | --kd-c-switch-inner-spacing-margin               | -                      | 0 5px 0 17px       |
|             | --kd-c-switch-inner-spacing-margin-checked       | -                      | 0 17px 0 5px       |
|             | --kd-c-switch-inner-spacing-margin-large         | -                      | 0 5px 0 21px       |
|             | --kd-c-switch-inner-spacing-margin-large-checked | -                      | 0 21px 0 5px       |
