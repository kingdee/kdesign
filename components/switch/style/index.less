@import '../../style/themes/index';
@import './mixin.less';

@switch-prefix-cls: ~'@{kd-prefix}-switch';
@icon-prefix-cls: ~'@{kd-prefix}icon';


.@{switch-prefix-cls} {
  .switch;

  .@{switch-prefix-cls}-inner {
    .switch-inner;
    user-select: none;
  }

  .@{switch-prefix-cls}-loading-icon {
    color: @disabled-color;
    position: absolute;
    box-sizing: border-box;
    display: inline-block;
    left: 1px;
    top: 50%;
    transform: translateY(-50%);
    z-index: 10;

    .@{icon-prefix-cls}-loadding-circle {
      animation: loadingcircle 1s infinite;
      animation-timing-function: cubic-bezier(0.42, 0.00, 0.58, 1.00);
      font-size: @switch-small-font-size;
      line-height: 1;
      vertical-align: baseline;
    }
  }

  &-disabled {
    background-color: @switch-off-disabled-bg;
    cursor: not-allowed;
  }

  &-checked {
    background-color: @switch-on-color;
    text-align: right;

    &::after {
      left: 100%;
      margin-left: -1px;
      transform: translateX(-100%);
    }

    .@{switch-prefix-cls}-loading-icon {
      left: unset;
      right: 1px;
      color: @switch-loading-color;
    }

    &.@{switch-prefix-cls}-disabled {
      background-color: @switch-on-disabled-bg;
    }

    .@{switch-prefix-cls}-inner {
      margin: @switch-inner-checked-margin;
    }

    &.@{switch-prefix-cls}-size-large {
      .@{switch-prefix-cls}-inner {
        margin: @switch-inner-large-checked-margin;
      }
    }
  }

  &-size-small {
    .switch-size(@switch-small-height, @switch-small-line-height, @switch-small-font-size, @switch-small-min-width, @switch-small-border-radius);
  }

  &-size-large {
    .switch-size(@switch-large-height, @switch-large-line-height, @switch-large-font-size, @switch-large-min-width, @switch-large-border-radius);

    .@{switch-prefix-cls}-inner {
      margin: @switch-inner-large-margin;
    }
  }
}
