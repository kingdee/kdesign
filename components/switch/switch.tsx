import React, { FC, useContext } from 'react'
import classNames from 'classnames'
import Icon from '../icon'
import ConfigContext from '../config-provider/ConfigContext'
import { getCompProps } from '../_utils'
import { tuple } from '../_utils/type'
import { useMergedState } from '../_utils/hooks'

// type SwitchSize = 'large' | 'middle' | 'small'
export const SwitchSizes = tuple('large', 'small')
export type SwitchSize = typeof SwitchSizes[number]
export type SwitchChangeEventHandler = (checked: boolean, event: MouseEvent) => void
export type SwitchClickEventHandler = SwitchChangeEventHandler
export interface ISwitchProps {
  size?: SwitchSize
  style?: Record<string, unknown> // 内联样式
  className?: string // 样式名
  checked?: boolean
  defaultChecked?: boolean
  disabled?: boolean
  loading?: boolean
  checkedChildren?: React.ReactNode
  unCheckedChildren?: React.ReactNode
  onClick?: SwitchClickEventHandler // 点击事件
  onChange?: SwitchChangeEventHandler // 变化事件
}

const Switch: FC<ISwitchProps> = (props) => {
  const { getPrefixCls, prefixCls, compDefaultProps: userDefaultProps, direction } = useContext(ConfigContext)
  const switchProps = getCompProps('Switch', userDefaultProps, props)
  const {
    size,
    className,
    prefixCls: customPrefixcls,
    checked,
    defaultChecked,
    checkedChildren,
    unCheckedChildren,
    loading,
    disabled,
    onClick,
    onChange,
    ...others
  } = switchProps

  const [newChecked, setChecked] = useMergedState(false, {
    value: checked,
    defaultValue: defaultChecked,
  })
  const switchPrefixCls = getPrefixCls!(prefixCls, 'switch', customPrefixcls)
  const rtlCls = direction === 'rtl' ? `${switchPrefixCls}-rtl` : null
  const handleClick = (e: React.MouseEvent<HTMLDivElement | HTMLAnchorElement, MouseEvent>) => {
    const newV = triggerChange(!newChecked, e)
    onClick && onClick(newV, e)
  }

  function triggerChange(newV: boolean, e: React.MouseEvent<HTMLDivElement | HTMLAnchorElement, MouseEvent>) {
    const isDisabled = disabled || loading
    let afterChangedChecked = newChecked
    if (!isDisabled) {
      afterChangedChecked = newV
      setChecked(afterChangedChecked)
      onChange?.(afterChangedChecked, e)
    }

    return afterChangedChecked
  }
  // ref
  const switchClasses = classNames(switchPrefixCls, rtlCls, className, {
    [`${switchPrefixCls}-size-${size}`]: size, // 尺寸样式
    [`${switchPrefixCls}-disabled`]: disabled || loading, // 禁用态样式
    [`${switchPrefixCls}-loading`]: loading, // 加载中样式
    [`${switchPrefixCls}-checked`]: newChecked, // 选中样式
  })

  const loadingIcon = loading ? (
    <span className={`${switchPrefixCls}-loading-icon`}>
      <Icon type="loadding-circle" spin />
    </span>
  ) : null

  return (
    <span className={switchClasses} {...others} onClick={handleClick}>
      <div className={`${switchPrefixCls}-inner`}>
        {loadingIcon}
        {newChecked ? checkedChildren : unCheckedChildren}
      </div>
    </span>
  )
}

Switch.displayName = 'Switch'
export default Switch
