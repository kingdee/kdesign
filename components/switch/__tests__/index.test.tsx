import React from 'react'
import { render, mount } from 'enzyme'
import mountTest from '../../../tests/shared/mountTest'
import Switch from '../index'
import { SwitchSizes } from '../switch'

describe('Switch', () => {
  // 1.mount test
  mountTest(Switch)

  // 2.render test
  it('renders correctly', () => {
    expect(render(<Switch />)).toMatchSnapshot()
    SwitchSizes.forEach((size) => {
      expect(render(<Switch size={size} />)).toMatchSnapshot()
    })
  })

  // 3.render no child without errors
  it('render no child without errors', () => {
    expect(mount(<Switch></Switch>)).toMatchSnapshot()
  })

  // 4.render null or undefined without errors
  it('render null or undefined without errors', () => {
    expect(
      mount(
        <Switch>
          {null}
          {undefined}
        </Switch>,
      ),
    ).toMatchSnapshot()
  })
  // 5.displayName
  it('should have displayName static property', () => {
    const wrapper = mount(<Switch />)
    expect((wrapper.type() as any).displayName).toBe('Switch')
  })
  // 6.className
  it('should class use right', () => {
    const DefaultSwitch = mount(<Switch />)
    expect(DefaultSwitch.find('.kd-switch')).toHaveClassName('.kd-switch-size-small')

    const DefaultCheckedSwitch = mount(<Switch defaultChecked />)
    expect(DefaultCheckedSwitch.find('.kd-switch')).toHaveClassName('.kd-switch-checked')

    const CheckedSwitch = mount(<Switch checked={true} />)
    expect(CheckedSwitch.find('.kd-switch')).toHaveClassName('.kd-switch-checked')

    const LoadingSwitch = mount(<Switch loading defaultChecked />)
    expect(LoadingSwitch.find('.kd-switch')).toHaveClassName('.kd-switch-loading')
    expect(LoadingSwitch.find('.kd-switch')).toHaveClassName('.kd-switch-disabled')

    const DisabledSwitch = mount(<Switch disabled />)
    expect(DisabledSwitch.find('.kd-switch')).toHaveClassName('.kd-switch-disabled')
  })
  // can not be clicked when disabled or loading
  it('should not clickable when switch is loading', () => {
    const onClick = jest.fn()
    const wrapper = mount(<Switch loading onClick={onClick} />)
    wrapper.simulate('click')
    expect(onClick).not.toHaveBeenCalledWith()
  })

  it('should not clickable when switch is disabled', () => {
    const onClick = jest.fn()
    const wrapper = mount(<Switch disabled onClick={onClick} />)
    wrapper.simulate('click')
    expect(onClick).not.toHaveBeenCalledWith()
  })

  const defaultSwitchProps = {
    unCheckedChildren: '关闭',
    checkedChildren: '开启',
    onClick: jest.fn(),
    onChange: jest.fn(),
  }
  // api
  it('api test', () => {
    const wrapper = mount(<Switch {...defaultSwitchProps}></Switch>)

    // unCheckedChildren and checkedChildren
    expect(wrapper.find('.kd-switch-inner')).toHaveText('关闭')
    wrapper.simulate('click')
    expect(wrapper.find('.kd-switch-inner')).toHaveText('开启')
    // onClick
    expect(defaultSwitchProps.onClick).toHaveBeenCalled()
    // onChange
    expect(defaultSwitchProps.onChange).toHaveBeenCalled()
  })
})
