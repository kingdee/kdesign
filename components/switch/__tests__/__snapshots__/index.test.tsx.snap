// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch render no child without errors 1`] = `
<Switch>
  <span
    className="kd-switch kd-switch-size-small"
    onClick={[Function]}
  >
    <div
      className="kd-switch-inner"
    />
  </span>
</Switch>
`;

exports[`Switch render null or undefined without errors 1`] = `
<Switch>
  <span
    className="kd-switch kd-switch-size-small"
    onClick={[Function]}
  >
    <div
      className="kd-switch-inner"
    />
  </span>
</Switch>
`;

exports[`Switch renders correctly 1`] = `
<span
  class="kd-switch kd-switch-size-small"
>
  <div
    class="kd-switch-inner"
  />
</span>
`;

exports[`Switch renders correctly 2`] = `
<span
  class="kd-switch kd-switch-size-large"
>
  <div
    class="kd-switch-inner"
  />
</span>
`;

exports[`Switch renders correctly 3`] = `
<span
  class="kd-switch kd-switch-size-small"
>
  <div
    class="kd-switch-inner"
  />
</span>
`;
